programs/survival/bucketstack.sc

```diff
@@ -0,0 +1,223 @@
+__config() ->
+(
+	m(
+		l('scope','player'),
+		l('stay_loaded',true),
+
+		l('commands',
+			m(
+				l('<bucket>','__print'),
+				l('<bucket> <stack>','__change')
+			)
+		),
+
+		l('arguments',
+			m(
+				l('bucket',
+					m(
+						l('type','term'),
+						l('options',filter(item_list(),_ ~ '_bucket') - '_bucket'),
+						l('suggest',filter(item_list(),_ ~ '_bucket') - '_bucket')
+					)
+				),
+				l('stack',
+					m(
+						l('type','int'),
+						l('min',1),
+						l('max',2^31-1),
+						l('suggest',l(1,2,4,8,16,32,64))
+					)
+				)
+			)
+		)
+	)
+);
+
+__on_player_collides_with_entity(player,entity) ->
+(
+	if(entity ~ 'pickup_delay' == 0,
+		//show some pretty particles when we pick an item up off the ground(delay == 0)
+		global_particle_bool = true;
+		__pickup_bucket_entity(player,entity)
+	);
+);
+
+__on_player_uses_item(player, item_tuple, hand) ->
+(
+	l(item,count,nbt) = item_tuple;
+	if(item == 'bucket',
+		bucket_item = query(player,'trace',4.5,'liquids') + '_bucket';
+		schedule(0,'__get_bucket_entity',player,bucket_item)
+	)
+);
+
+__on_player_interacts_with_entity(player, entity, hand) ->
+(
+	//only check everything else after we know that the player has no empty slots
+	//otherwise the vanilla behavior will do the hard work for us
+	if(inventory_find(player,null) > 35,
+		if(hand == 'mainhand',
+			slot = player~'selected_slot',
+			//-1 is offhand
+			slot = -1
+		);
+		item = inventory_get(player,slot):0;
+		bucket_prefix = lower(replace(entity,' ','_'));
+		if(item == 'bucket' && (entity == 'Cow' || entity == 'Goat'),
+			product_item = 'milk_bucket'
+		);
+		//lower because entity names are capitalized, bucket names are lowercase
+		//replace for tropical fish
+		if(item == 'water_bucket' && keys(load_app_data()) ~ bucket_prefix,
+			product_item = bucket_prefix + '_bucket'
+		);
+		//use schedule() because item entity doesn't exist until end of tick
+		if(product_item,
+			schedule(0,'__get_bucket_entity',player,product_item)
+		)
+	)
+);
+
+__initialize() ->
+(
+	if(!load_app_data(),
+		buckets = m();
+		for(filter(item_list(),_ ~ '_bucket') - '_bucket',
+			buckets:_ = 1;
+		);
+		store_app_data(buckets);
+	)
+);

```

Comment:
I'm not sure about removnig the in-app configuration of default stack sizes. It makes it easier to use, imo (just edit the values when savign the app, no need to run eight different commands in-game). I'd say, leave the in-app configured values and override them with the command defined ones, if given.

Reply:
It makes it easier only for first time use and only if you want to change many of the stack sizes at once. I don't know what you think, but I suspect that most people would change water, lava, powder snow, and maybe milk or axolotls. I don't think that many people would want to change every single stack size, so only a few commands would be required. Matching tropical fish nbt is very hard and unless you need a lot of the other fish, there's no need to change the default stack size.

I can put the configuration map back in, but I would make sure that the app data values override it.

Reply:
Tropical fish buckets are used to breed axolotl, for instance. And I don't think it would be unresonable to think people would want all buckets to stack to the same ammount. That being said, It's enough for me with the `all` command added and, ultimately, it's your app. I'm not gonna block it because it's not the way I'd do it. When I can, I'll test it in game and approve if I don't find bugs :)

Reply:
I can understand wanting tropical fish buckets to stack, but I'm only allowing that when nbt matches because otherwise players will be able to dupe any fish type.

Reply:
Yes, yes, I agree with that.